src/main/kotlin/com/openlattice/data/storage/PostgresEntityDataQueryService.kt

```diff
@@ -973,31 +983,39 @@ class PostgresEntityDataQueryService(
      * @param partitions Contains the partition information for the requested entity set.
      */
     private fun tombstoneIdsTable(
-            conn: Connection,
             entitySetId: UUID,
             entityKeyIds: Set<UUID>,
             version: Long,
             partitions: List<Int> = partitionManager.getEntitySetPartitions(entitySetId).toList()
     ): WriteEvent {
-        val entityKeyIdsArr = PostgresArrays.createUuidArray(conn, entityKeyIds)
+        return hds.connection.use { conn ->
+            conn.autoCommit = false
+            val partitionsMap = entityKeyIds.groupBy { getPartition(it, partitions) }
 
-        val partitionsArr = PostgresArrays.createIntArray(conn, entityKeyIds.map {
-            getPartition(
-                    it, partitions
-            )
-        })
+            val lockedEntities = conn.prepareStatement(bulkLockEntitiesInIdsTable)
 
-        val numUpdated = conn.prepareStatement(updateVersionsForEntitiesInEntitySet).use { ps ->
-            ps.setLong(1, -version)
-            ps.setLong(2, -version)
-            ps.setLong(3, -version)
-            ps.setObject(4, entitySetId)
-            ps.setArray(5, entityKeyIdsArr)
-            ps.setArray(6, partitionsArr)
-            ps.executeUpdate()
-        }
+            partitionsMap.forEach { (partition, ids) ->
+                lockedEntities.setArray(2, PostgresArrays.createUuidArray(conn, ids))
+                lockedEntities.setInt(3, partition)
+                lockedEntities.addBatch()
+            }
+            lockedEntities.executeBatch()
 
-        return WriteEvent(version, numUpdated)
+            val entityKeyIdsArr = PostgresArrays.createUuidArray(conn, entityKeyIds)
+            val partitionsArr = PostgresArrays.createIntArray(conn, partitionsMap.keys)
+
+            val numUpdated = conn.prepareStatement(updateVersionsForEntitiesInEntitySet).use { ps ->
+                ps.setLong(1, -version)
+                ps.setLong(2, -version)
+                ps.setLong(3, -version)
+                ps.setObject(4, entitySetId)
+                ps.setArray(5, entityKeyIdsArr)
+                ps.setArray(6, partitionsArr)
```

Comment:
just a note that I think my PR changed the bind order of these last 2 params, so that's something to look out for on merging these

Reply:
:+1: